pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/LookupTransformFunction.java

```diff
@@ -0,0 +1,332 @@
+    _dataManager = DimensionTableDataManager.getInstanceByTableName(_dimTableName);
+    Preconditions.checkArgument(_dataManager != null,
+            String.format("Dimension table does not exist: %s", _dimTableName));
+
+    _lookupColumnFieldSpec = _dataManager.getColumnFieldSpec(_dimColumnName);
+    Preconditions.checkArgument(_lookupColumnFieldSpec != null,
+        String.format("Column does not exist in dimension table: %s:%s", _dimTableName, _dimColumnName));
+
+    for (String joinKey : _joinKeys) {
+      FieldSpec pkColumnSpec = _dataManager.getColumnFieldSpec(joinKey);
+      Preconditions.checkArgument(pkColumnSpec != null,
+          String.format("Primary key column doesn't exist in dimension table: %s:%s", _dimTableName, joinKey));
+      _joinValueFieldSpecs.add(pkColumnSpec);
+    }
+
+    List<String> tablePrimaryKeyColumns = _dataManager.getPrimaryKeyColumns();
+    Preconditions.checkArgument(_joinKeys.equals(tablePrimaryKeyColumns),
+        String.format("Provided join keys (%s) must be the same as table primary keys: %s", _joinKeys,
+            tablePrimaryKeyColumns));
+  }
+
+  @Override
+  public TransformResultMetadata getResultMetadata() {
+    return new TransformResultMetadata(_lookupColumnFieldSpec.getDataType(),
+        _lookupColumnFieldSpec.isSingleValueField(), false);
+  }
+
+  private Object[] lookup(ProjectionBlock projectionBlock) {
+    int numPkColumns = _joinKeys.size();
+    int numDocuments = projectionBlock.getNumDocs();
+    Object[][] pkColumns = new Object[numPkColumns][];
+    for (int i = 0; i < numPkColumns; i++) {
+      FieldSpec.DataType colType = _joinValueFieldSpecs.get(i).getDataType();
+      TransformFunction tf = _joinValueFunctions.get(i);
+      switch (colType) {
+        case STRING:
+          pkColumns[i] = tf.transformToStringValuesSV(projectionBlock);
+          break;
+        case INT:
+          pkColumns[i] = ArrayUtils.toObject(tf.transformToIntValuesSV(projectionBlock));
+          break;
+        case LONG:
+          pkColumns[i] = ArrayUtils.toObject(tf.transformToLongValuesSV(projectionBlock));
+          break;
+        case FLOAT:
+          pkColumns[i] = ArrayUtils.toObject(tf.transformToFloatValuesSV(projectionBlock));
+          break;
+        case DOUBLE:
+          pkColumns[i] = ArrayUtils.toObject(tf.transformToDoubleValuesSV(projectionBlock));
+          break;
+        case BYTES:
+          byte[][] primitiveValues = tf.transformToBytesValuesSV(projectionBlock);
+          pkColumns[i] = new Byte[numDocuments][];
+          for (int n = 0; n < numDocuments; n++) {
+            pkColumns[i][n] = ArrayUtils.toObject(primitiveValues[n]);
+          }
+          break;
+        default:
+          throw new IllegalStateException("Unknown column type for primary key");
+      }
+    }
+
+    Object[] resultSet = new Object[numDocuments];
+    for (int i = 0; i < numDocuments; i++) {
+      // prepare pk
+      Object[] pkValues = new Object[numPkColumns];
```

Comment:
Reuse this `Object[]` instead of creating a new one per iteration

Reply:
I can do something like:
```java
  Object[] resultSet = new Object[numDocuments];
  Object[] pkValues = new Object[numPkColumns];
  for (int i = 0; i < numDocuments; i++) {
    // prepare pk
    for (int j = 0; j < numPkColumns; j++) {
      pkValues[j] = pkColumns[j][i];
    }
    // lookup
    GenericRow row = _dataManager.lookupRowByPrimaryKey(new PrimaryKey(pkValues));
    if (row != null) {
      resultSet[i] = row.getValue(_dimColumnName);
    }
  }
```
I don't see much point in doing the same for `PrimaryKey` though. We will only be reusing the pointer which is not really helpful.

Reply:
It will reuse the object instead of creating one per doc:

```
  Object[] resultSet = new Object[numDocuments];
  Object[] pkValues = new Object[numPkColumns];
  PrimaryKey primaryKey = new PrimaryKey(pkValues);
  for (int i = 0; i < numDocuments; i++) {
    ...
```

Reply:
I don't think this will work. We have to create a new 'PrimaryKey' instance per document since the values (`pkValues`) are going to be different.

Maybe I should have created a `PrimaryKey[]` in the beginning of this method, rather than a `pkColumn[]` which probably confused you. Let me look into cleaning this up a bit.

Reply:
You can directly modify `pkValues` without changing `primaryKey`. They share the same reference.
This is not critical, so both way is fine. It just save minor garbages